File: src/guide/models/encounter.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from .shared import Condition, EncounterStatus


class ActionBudget(BaseModel):
    has_action: bool = True
    has_bonus_action: bool = True
    has_reaction: bool = True
    movement_remaining: int = 30

    @classmethod
    def new(cls, speed: int = 30) -> "ActionBudget":
        return cls(movement_remaining=speed)

    def reset(self, speed: int = 30) -> None:
        self.has_action = True
        self.has_bonus_action = True
        self.has_reaction = True
        self.movement_remaining = speed


class CombatParticipant(BaseModel):
    id: UUID
    encounter_id: UUID
    character_id: UUID
    name: str
    initiative_roll: int = 0
    initiative_modifier: int = 0
    initiative_total: int = 0
    current_hp: int = 10
    max_hp: int = 10
    armor_class: int = 10
    conditions: list[Condition] = Field(default_factory=list)
    action_budget: ActionBudget = Field(default_factory=ActionBudget)
    has_taken_turn: bool = False
    is_defeated: bool = False


class Encounter(BaseModel):
    id: UUID
    session_id: UUID
    campaign_id: UUID
    name: str | None = None
    description: str | None = None
    status: EncounterStatus = EncounterStatus.pending
    round: int = 0
    current_turn_index: int = 0
    participants: list[CombatParticipant] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class EncounterSummary(BaseModel):
    encounter: Encounter
    current_participant: CombatParticipant | None = None
    round: int


class CreateEncounterRequest(BaseModel):
    session_id: UUID
    name: str | None = None
    description: str | None = None
    participant_character_ids: list[UUID] = Field(default_factory=list)


class AddParticipantRequest(BaseModel):
    character_id: UUID
    initiative_roll: int | None = None


class UpdateParticipantRequest(BaseModel):
    hp_delta: int | None = None
    set_hp: int | None = None
    add_condition: Condition | None = None
    remove_condition: Condition | None = None
    spend_action: bool | None = None
    spend_bonus_action: bool | None = None
    spend_reaction: bool | None = None
    spend_movement: int | None = None

File: crates/guide-core/src/models/encounter.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::shared::{Condition, EncounterStatus};

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct Encounter {
    pub id: Uuid,
    pub session_id: Uuid,
    pub campaign_id: Uuid,
    pub name: Option<String>,
    pub description: Option<String>,
    pub status: EncounterStatus,
    pub round: i32,
    pub current_turn_index: i32,
    pub participants: Vec<CombatParticipant>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct CombatParticipant {
    pub id: Uuid,
    pub encounter_id: Uuid,
    pub character_id: Uuid,
    pub name: String,
    pub initiative_roll: i32,
    pub initiative_modifier: i32,
    pub initiative_total: i32,
    pub current_hp: i32,
    pub max_hp: i32,
    pub armor_class: i32,
    pub conditions: Vec<Condition>,
    pub action_budget: ActionBudget,
    pub has_taken_turn: bool,
    pub is_defeated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct ActionBudget {
    pub has_action: bool,
    pub has_bonus_action: bool,
    pub has_reaction: bool,
    pub movement_remaining: i32,
}

impl ActionBudget {
    pub fn new(speed: i32) -> Self {
        Self {
            has_action: true,
            has_bonus_action: true,
            has_reaction: true,
            movement_remaining: speed,
        }
    }

    pub fn reset(&mut self, speed: i32) {
        self.has_action = true;
        self.has_bonus_action = true;
        self.has_reaction = true;
        self.movement_remaining = speed;
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct CreateEncounterRequest {
    pub session_id: Uuid,
    pub name: Option<String>,
    pub description: Option<String>,
    pub participant_character_ids: Vec<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct AddParticipantRequest {
    pub character_id: Uuid,
    pub initiative_roll: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct UpdateParticipantRequest {
    pub hp_delta: Option<i32>,
    pub set_hp: Option<i32>,
    pub add_condition: Option<Condition>,
    pub remove_condition: Option<Condition>,
    pub spend_action: Option<bool>,
    pub spend_bonus_action: Option<bool>,
    pub spend_reaction: Option<bool>,
    pub spend_movement: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct EncounterSummary {
    pub encounter: Encounter,
    pub current_participant: Option<CombatParticipant>,
    pub round: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct GeneratedEncounter {
    pub title: String,
    pub description: String,
    pub encounter_type: GeneratedEncounterType,
    pub challenge_rating: Option<f32>,
    pub suggested_enemies: Vec<EnemySuggestion>,
    pub narrative_hook: String,
    pub alternative: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum GeneratedEncounterType {
    Combat,
    Social,
    Exploration,
    Puzzle,
    Mixed,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
pub struct EnemySuggestion {
    pub name: String,
    pub count: u32,
    pub cr: Option<f32>,
}
